server/src/main/java/org/opensearch/search/internal/ContextIndexSearcher.java

```diff
@@ -310,17 +312,25 @@ private void searchLeaf(LeafReaderContext ctx, Weight weight, Collector collecto
             return;
         }
 
-        cancellable.checkCancelled();
-        weight = wrapWeight(weight);
-        // See please https://github.com/apache/lucene/pull/964
-        collector.setWeight(weight);
         final LeafCollector leafCollector;
         try {
+            cancellable.checkCancelled();
+            weight = wrapWeight(weight);
+            // See please https://github.com/apache/lucene/pull/964
+            collector.setWeight(weight);
             leafCollector = collector.getLeafCollector(ctx);
         } catch (CollectionTerminatedException e) {
             // there is no doc of interest in this reader context
             // continue with the following leaf
             return;
+        } catch (QueryPhase.TimeExceededException e) {
+            // Swallow timeout exception and return
+            if (searchContext.request().allowPartialSearchResults() == false) {
+                // Can't rethrow TimeExceededException because not serializable
+                throw new QueryPhaseExecutionException(searchContext.shardTarget(), "Time exceeded");
```

Comment:
This is my suggestion, let me know what you think:
1. It seems like there is some gap in `ConcurrentQueryPhaseSearcher::searchWithCollectorManager` exception handling as from lucene it will always receive a `RunTimeException` wrapping an `ExecutionException` wrapping the original exception thrown while search for concurrent case. We will need to handle this `RuntimeExecption` and wrapped `ExecutionException` [here](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java#L83) correctly and still be able to throw the `TimeExceededException` from here ? Right now I don't see any code path throwing either of the caught exception. 
2. I think we should keep the behavior same irrespective of `partial` results set to true or false. Where it consumes the exception and waits for all the slices to complete. That will be fine in most of the cases, since other slice thread will also immediately return if one threads hits the cancellation except when there is lot of contention in the queue. IMO, Keeping it same will remove the bimodal behavior and will help to reason about things easily. Once lucene timeout changes are merge in we can see if we can somehow improve on the timeout handling for concurrent path.
3. Once all the slices are completed, they will get chance to `reduce` which in `timeout` case either none of the collector collected anything or some collectors performed the collection and eventually perform the reduce on the collected results. Post reduce phase based on the state in `SearchContext` and if partial result is requested or not, we can handle setting the proper flag in `QueryResult` [here](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/search/query/ConcurrentQueryPhaseSearcher.java#L82) or throw corresponding exception. My preference is to update the `QueryResult` in single place which is `QueryPhaseSearcher` instead of here.

Reply:
@sohami Thanks for the suggestions! I see now that the exception handling in my PR doesn't quite work for the `allowPartialSearchResults == false` case because a different exception (`RuntimeException`) is going to be thrown. I do think this gets us to the same situation, the shard search failure is just for a different reason now. However, I do agree with the observations in point 2 and I think simplifying the exception handling in general is a good idea so I will work on making these changes.